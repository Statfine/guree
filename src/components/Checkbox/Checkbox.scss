@import "../../common/utils/index.scss";

$checkbox: scopedClass(checkbox);

.#{$checkbox} {
  line-height: 1.5em;
  @include inline-flex-start-center;
  cursor: pointer;
  &-inner {
    @include inline-flex-center-center;
    width: 16px;
    height: 16px;
    border: 1px solid #ccc;
    margin-right: 4px;
  }
  &-name {
    line-height: 1.5em;
  }
  & + & {
    margin-left: 20px;
  }

  &-checked {
    .#{$checkbox}-inner {
      background-color: #409eff;
    }
  }

  &-indeterminate {
    .#{$checkbox}-inner {
      border: 1px solid #409eff;
      &:after {
        content: "";
        width: 8px;
        height: 8px;
        background-color: #409eff;
        display: inline-block;
      }
    }
  }

  &-disabled {
    cursor: not-allowed;
    color: rgba($color: #000000, $alpha: .25);
    &.#{$checkbox}-checked{
      .#{$checkbox}-inner {
        background-color: rgba($color: #000000, $alpha: .25);
      }
    }
    .#{$checkbox}-inner {
      border: 1px solid rgba($color: #000000, $alpha: .25);
      &:after {
        background-color: rgba($color: #000000, $alpha: .25);
      }
    }
  }

  &-group {
    @include inline-flex-start-center;
  }
}