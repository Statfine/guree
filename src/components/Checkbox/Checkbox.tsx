import React, { ReactNode, useState, MouseEventHandler, useEffect, useRef } from "react";
import "./Checkbox.scss";
import {
  combineClasses,
  scopedClass,
  classesObj,
  updateStateOnPropChange,
} from "@utils/index";
import Icon from "@components/Icon";

export interface checkAttrs {
  children: ReactNode;
  value: any;
  disabled: boolean;
  checked: boolean;
  indeterminate: boolean;
}

interface Props {
  children?: ReactNode;
  /** Checkbox 的值，可用于与其他 Checkbox 区分 */
  value?: any;
  /** 选中状态 */
  checked?: boolean;
  /** 禁用状态 */
  disabled?: boolean;
  /** 设置 indeterminate 状态，只负责样式控制 */
  indeterminate?: boolean;
  /** 点击时调用 */
  onClick?: (checkAttrs: checkAttrs) => {};
}

const Checkbox: React.FunctionComponent<Props> = ({
  children,
  value,
  checked = false,
  disabled = false,
  indeterminate = false,
  onClick,
}: Props) => {
  const checkboxClass = (...classes: (string | Array<string> | classesObj)[]) =>
    scopedClass("checkbox", ...classes);

  const [checkedStatus, setCheckedStatus] = useState(checked);
  const check = () => {
    setCheckedStatus(!checkedStatus);
  };

  const checkboxClassName = combineClasses({
    [checkboxClass()]: true,
    [checkboxClass("checked")]: checkedStatus,
    [checkboxClass("disabled")]: disabled,
    [checkboxClass("indeterminate")]: !checkedStatus && indeterminate,
  });

  const clickHandler: MouseEventHandler<HTMLElement> = (e) => {
    if (disabled) {
      return;
    }
    check();
    onClick &&
      onClick({
        children,
        value,
        disabled,
        checked: !checkedStatus,
        indeterminate,
      });
  };

  updateStateOnPropChange(checked, setCheckedStatus);



  return (
    <span className={checkboxClassName} onClick={clickHandler}>
      <span className={checkboxClass("inner")}>
        {checkedStatus ? (
          <Icon name="check" fill="#fff" size="10px"></Icon>
        ) : null}
      </span>
      <span className={checkboxClass("name")}>{children}</span>
    </span>
  );
};

export default Checkbox;
