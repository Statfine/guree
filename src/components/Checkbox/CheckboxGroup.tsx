import React, { useState, ReactElement, useRef, useEffect } from "react";
import { checkAttrs } from "./Checkbox";
import "./Checkbox.scss";
import { scopedClass, classesObj, updateStateOnPropChange } from "@utils/index";
import { isUndefined } from "util";

interface Props {
  children: ReactElement;
  value?: Array<any>;
  onChange?: (checkedArr?: Array<any>) => {};
  defaultValue?: Array<any>;
}

const CheckGroup: React.FunctionComponent<Props> = ({
  children,
  value,
  onChange,
  defaultValue,
}: Props) => {
  const radioGroupClass = (
    ...classes: (string | Array<string> | classesObj)[]
  ) => scopedClass("radio-group", ...classes);

  const [checkedValues, setCheckedValues] = useState(
    isUndefined(value) ? defaultValue : value
  );

  const changeChecked = (checkAttrs: checkAttrs) => {
    React.Children.forEach(children, (element) => {
      const { value: elementValue } = element.props;
      const emptyArr: Array<any> = [];
      if (elementValue === checkAttrs.value) {
        const values = checkedValues
          ? checkAttrs.checked
            ? checkedValues.concat([checkAttrs.value])
            : checkedValues.filter((v: any) => v !== checkAttrs.value)
          : emptyArr.concat([checkAttrs.value]);
        onChange && onChange(values);
        if (!isUndefined(value)) {
          return;
        }
        setCheckedValues(values);
      }
    });
  };
  updateStateOnPropChange(value, setCheckedValues);

  return (
    <div className={radioGroupClass()}>
      {React.Children.map(children, (element) => {
        return React.cloneElement(
          element,
          Object.assign(
            {},
            checkedValues
              ? {
                  checked: checkedValues.includes(element.props.value),
                  onChange: changeChecked,
                }
              : {
                  onChange: changeChecked,
                },
            element.props
          )
        );
      })}
    </div>
  );
};

export default CheckGroup;
