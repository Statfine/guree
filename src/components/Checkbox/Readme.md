### 基本用法

```jsx
const onChange = (checkboxAttrs) => {
  console.log(checkboxAttrs);
};
<Checkbox value={"checkbox"} onChange={onChange}>
  checkbox
</Checkbox>;
```

### 默认选中

```jsx
<Checkbox defaultChecked value={"radio"}>
  radio
</Checkbox>
```

### 受控组件

```jsx
import { useState } from "react";
const [checked, setChecked] = useState(false);
const onChange = (checkboxAttrs) => {
  const { checked } = checkboxAttrs;
  setChecked(checked);
};
<div>
  <Checkbox checked={checked} onChange={onChange} value={"checkbox"}>
    checkbox
  </Checkbox>
</div>;
```

### 禁用状态

```jsx
<Checkbox disabled value={"checkbox"}>
  checkbox
</Checkbox>
```

### Checkbox 组

```jsx
import { CheckboxGroup } from "@components/Checkbox";
<CheckboxGroup>
  <Checkbox value={1}>checkbox1</Checkbox>
  <Checkbox value={2}>checkbox2</Checkbox>
  <Checkbox value={3}>checkbox3</Checkbox>
</CheckboxGroup>;
```

### Checkbox 组的默认值

```jsx
import { CheckboxGroup } from "@components/Checkbox";
<CheckboxGroup defaultValue={[1, 2]}>
  <Checkbox value={1}>checkbox1</Checkbox>
  <Checkbox value={2}>checkbox2</Checkbox>
  <Checkbox value={3}>checkbox3</Checkbox>
</CheckboxGroup>;
```

### 受控的 Checkbox 组组件

```jsx
import { CheckboxGroup } from "@components/Checkbox";
import { useState } from "react";
const [checkedList, setCheckedList] = useState([]);
const onChange = (list) => {
  setCheckedList(list)
};
  <CheckboxGroup onChange={onChange} value={checkedList}>
    <Checkbox value={1}>checkbox1</Checkbox>
    <Checkbox value={2}>checkbox2</Checkbox>
    <Checkbox value={3}>checkbox3</Checkbox>
  </CheckboxGroup>
```

### 全选，需要用到 indeterminate 属性

```jsx
import { CheckboxGroup } from "@components/Checkbox";
import { useState } from "react";
const valueList = [1, 2, 3];
const [indeterminate, setIndeterminate] = useState(false);
const [checkedList, setCheckedList] = useState([]);
const [checkedAll, setCheckedAll] = useState(false);
const checkAll = (e) => {
  const { checked } = e;
  setCheckedAll(checked);
  if (checked) {
    setCheckedList(valueList);
  } else {
    setCheckedList([]);
  }
};
const onChange = (list) => {
  setCheckedList(list)
  setIndeterminate(list.length !== 0);
  setCheckedAll(list.length === valueList.length);
};
<div>
  <div>
    <Checkbox
      value={"checkAll"}
      onChange={checkAll}
      indeterminate={indeterminate}
      checked={checkedAll}
    >
      checkAll
    </Checkbox>
  </div>
  <CheckboxGroup onChange={onChange} value={checkedList}>
    <Checkbox value={valueList[0]}>checkbox1</Checkbox>
    <Checkbox value={valueList[1]}>checkbox2</Checkbox>
    <Checkbox value={valueList[2]}>checkbox3</Checkbox>
  </CheckboxGroup>
</div>;
```
