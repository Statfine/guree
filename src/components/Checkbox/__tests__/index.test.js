import React from "react";
import renderer from "react-test-renderer";
import Checkbox, { CheckboxGroup } from "..";
import { mount, shallow } from "enzyme";

describe("Checkbox 组件", () => {
  it("正确导出 Checkbox 组件", () => {
    const checkbox = renderer.create(<Checkbox></Checkbox>).toJSON();
    expect(checkbox).toMatchSnapshot();
  });

  it("测试 Checkbox 组件 value 参数", () => {
    const checkbox = renderer
      .create(<Checkbox value={"checkbox"}></Checkbox>)
      .toJSON();
    expect(checkbox).toMatchSnapshot();
  });

  it("测试 Checkbox 组件 checked 参数", () => {
    const checkbox = renderer.create(<Checkbox checked></Checkbox>).toJSON();
    expect(checkbox).toMatchSnapshot();
  });

  it("测试 Checkbox 组件 defaultValue 参数", () => {
    const checkbox = renderer.create(<Checkbox defaultValue></Checkbox>).toJSON();
    expect(checkbox).toMatchSnapshot();
  });

  it("测试 Checkbox 组件 disabled 参数", () => {
    const checkbox = renderer.create(<Checkbox disabled></Checkbox>).toJSON();
    expect(checkbox).toMatchSnapshot();
  });

  it("测试 Checkbox 组件 checked 和 defaultValue 并存", () => {
    const checkbox = renderer.create(<Checkbox checked defaultValue={false}></Checkbox>).toJSON();
    expect(checkbox).toMatchSnapshot();
  });

  it("测试 Checkbox 组件 indeterminate 参数", () => {
    const checkbox = renderer
      .create(<Checkbox indeterminate></Checkbox>)
      .toJSON();
    expect(checkbox).toMatchSnapshot();
  });

  it("测试 Checkbox 组件 indeterminate 和 checked 并存", () => {
    const checkbox = renderer
      .create(<Checkbox indeterminate checked></Checkbox>)
      .toJSON();
    expect(checkbox).toMatchSnapshot();
  });

  it("测试 Checkbox 组件 onChange 参数", () => {
    var value = null;
    const fn = (v) => (value = v.checked);
    const checkbox = mount(<Checkbox onChange={fn}></Checkbox>);
    
    checkbox.simulate("click");
    expect(value).toEqual(true);
    checkbox.update();
    expect(
      checkbox.find(".guree-checkbox").hasClass("guree-checkbox-checked")
    ).toEqual(true);

    checkbox.simulate("click");
    expect(value).toEqual(false);
    checkbox.update();
    expect(
      checkbox.find(".guree-checkbox").hasClass("guree-checkbox-checked")
    ).toEqual(false);
  });

  it("测试 Checkbox 组件 受控组件", () => {
    var value = null;
    const fn = (v) => (value = v.checked);
    const checkbox = mount(<Checkbox onChange={fn} checked={false}></Checkbox>);
    
    checkbox.simulate("click");
    expect(value).toEqual(true);
    checkbox.setProps({ checked: true });
    checkbox.update();
    expect(
      checkbox.find(".guree-checkbox").hasClass("guree-checkbox-checked")
    ).toEqual(true);

    checkbox.simulate("click");
    expect(value).toEqual(false);
    checkbox.setProps({ checked: false });
    checkbox.update();
    expect(
      checkbox.find(".guree-checkbox").hasClass("guree-checkbox-checked")
    ).toEqual(false);
  });

  it("测试 Checkbox 组件 disabled 状态下的点击事件", () => {
    const fn = jest.fn();
    const checkbox = mount(<Checkbox onChange={fn} disabled></Checkbox>);
    checkbox.simulate("click");
    expect(fn).not.toBeCalled();
    checkbox.update();
    expect(
      checkbox.find(".guree-checkbox").hasClass("guree-checkbox-checked")
    ).toEqual(false);
  });
});

describe("CheckboxGroup 组件", () => {
  it("正确导出 CheckboxGroup 组件", () => {
    const checkboxGroup = renderer
      .create(<CheckboxGroup></CheckboxGroup>)
      .toJSON();
    expect(checkboxGroup).toMatchSnapshot();
  });

  it("正确导出 CheckboxGroup 组件 value 参数", () => {
    const checkboxGroup = renderer
      .create(
        <CheckboxGroup value={[1]}>
          <Checkbox value={1}>1</Checkbox>
          <Checkbox value={2}>2</Checkbox>
          <Checkbox value={3}>3</Checkbox>
        </CheckboxGroup>
      )
      .toJSON();
    expect(checkboxGroup).toMatchSnapshot();
  });

  it("正确导出 CheckboxGroup 组件 defaultValue 参数", () => {
    const checkboxGroup = renderer
      .create(
        <CheckboxGroup defaultValue={[1]}>
          <Checkbox value={1}>1</Checkbox>
          <Checkbox value={2}>2</Checkbox>
          <Checkbox value={3}>3</Checkbox>
        </CheckboxGroup>
      )
      .toJSON();
    expect(checkboxGroup).toMatchSnapshot();
  });

  it("正确导出 CheckboxGroup 组件 onChange 参数", () => {
    const fn = jest.fn();
    const checkboxGroup = mount(
      <CheckboxGroup onChange={fn}>
        <Checkbox value={1}>1</Checkbox>
        <Checkbox value={2}>2</Checkbox>
        <Checkbox value={3}>3</Checkbox>
      </CheckboxGroup>
    );
    checkboxGroup.find(".guree-checkbox").forEach((node, index) => {
      if (index === 0 || index === 2) {
        node.simulate("click");
      }
    });
    expect(fn).toBeCalledWith([1, 3]);
    checkboxGroup.update();
    checkboxGroup.find(".guree-checkbox").forEach((node, index) => {
      expect(node.hasClass("guree-checkbox-checked")).toEqual(
        index === 0 || index === 2
      );
    });

    checkboxGroup.find(".guree-checkbox").forEach((node, index) => {
      if (index === 0 || index === 2) {
        node.simulate("click");
      }
    });
    expect(fn).toBeCalledWith([]);
    checkboxGroup.update();
    checkboxGroup.find(".guree-checkbox").forEach((node, index) => {
      expect(node.hasClass("guree-checkbox-checked")).toEqual(false);
    });
  });

  it("正确导出 CheckboxGroup 组件 受控组件", () => {
    const fn = jest.fn();
    const checkboxGroup = mount(
      <CheckboxGroup onChange={fn} value={[1]}>
        <Checkbox value={1}>1</Checkbox>
        <Checkbox value={2}>2</Checkbox>
        <Checkbox value={3}>3</Checkbox>
      </CheckboxGroup>
    );
    checkboxGroup.find(".guree-checkbox").forEach((node, index) => {
      if (index === 0 || index === 2) {
        node.simulate("click");
      }
    });
    expect(fn).toBeCalledWith([1, 3]);
    checkboxGroup.setProps({ value: [1, 3] });
    checkboxGroup.update();
    checkboxGroup.find(".guree-checkbox").forEach((node, index) => {
      expect(node.hasClass("guree-checkbox-checked")).toEqual(
        index === 0 || index === 2
      );
    });

    checkboxGroup.find(".guree-checkbox").forEach((node, index) => {
      if (index === 0 || index === 2) {
        node.simulate("click");
      }
    });
    expect(fn).toBeCalledWith([]);
    checkboxGroup.setProps({ value: [] });
    checkboxGroup.update();
    checkboxGroup.find(".guree-checkbox").forEach((node, index) => {
      expect(node.hasClass("guree-checkbox-checked")).toEqual(false);
    });
  });
});
