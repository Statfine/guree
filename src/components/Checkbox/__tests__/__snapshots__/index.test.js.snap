// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox 组件 正确导出 Checkbox 组件 1`] = `
<span
  className="guree-checkbox"
  onClick={[Function]}
>
  <span
    className="guree-checkbox-inner"
  />
  <span
    className="guree-checkbox-name"
  />
</span>
`;

exports[`Checkbox 组件 测试 Checkbox 组件 checked 参数 1`] = `
<span
  className="guree-checkbox guree-checkbox-checked"
  onClick={[Function]}
>
  <span
    className="guree-checkbox-inner"
  >
    <svg
      className="guree-icon"
      style={
        Object {
          "fill": "#fff",
          "height": "10px",
          "width": "10px",
        }
      }
    >
      <use
        xlinkHref="#check"
      />
    </svg>
  </span>
  <span
    className="guree-checkbox-name"
  />
</span>
`;

exports[`Checkbox 组件 测试 Checkbox 组件 checked 和 defaultValue 并存 1`] = `
<span
  className="guree-checkbox guree-checkbox-checked"
  onClick={[Function]}
>
  <span
    className="guree-checkbox-inner"
  >
    <svg
      className="guree-icon"
      style={
        Object {
          "fill": "#fff",
          "height": "10px",
          "width": "10px",
        }
      }
    >
      <use
        xlinkHref="#check"
      />
    </svg>
  </span>
  <span
    className="guree-checkbox-name"
  />
</span>
`;

exports[`Checkbox 组件 测试 Checkbox 组件 defaultValue 参数 1`] = `
<span
  className="guree-checkbox"
  onClick={[Function]}
>
  <span
    className="guree-checkbox-inner"
  />
  <span
    className="guree-checkbox-name"
  />
</span>
`;

exports[`Checkbox 组件 测试 Checkbox 组件 disabled 参数 1`] = `
<span
  className="guree-checkbox guree-checkbox-disabled"
  onClick={[Function]}
>
  <span
    className="guree-checkbox-inner"
  />
  <span
    className="guree-checkbox-name"
  />
</span>
`;

exports[`Checkbox 组件 测试 Checkbox 组件 indeterminate 参数 1`] = `
<span
  className="guree-checkbox guree-checkbox-indeterminate"
  onClick={[Function]}
>
  <span
    className="guree-checkbox-inner"
  />
  <span
    className="guree-checkbox-name"
  />
</span>
`;

exports[`Checkbox 组件 测试 Checkbox 组件 indeterminate 和 checked 并存 1`] = `
<span
  className="guree-checkbox guree-checkbox-checked"
  onClick={[Function]}
>
  <span
    className="guree-checkbox-inner"
  >
    <svg
      className="guree-icon"
      style={
        Object {
          "fill": "#fff",
          "height": "10px",
          "width": "10px",
        }
      }
    >
      <use
        xlinkHref="#check"
      />
    </svg>
  </span>
  <span
    className="guree-checkbox-name"
  />
</span>
`;

exports[`Checkbox 组件 测试 Checkbox 组件 value 参数 1`] = `
<span
  className="guree-checkbox"
  onClick={[Function]}
>
  <span
    className="guree-checkbox-inner"
  />
  <span
    className="guree-checkbox-name"
  />
</span>
`;

exports[`CheckboxGroup 组件 正确导出 CheckboxGroup 组件 1`] = `
<div
  className="guree-radio-group"
/>
`;

exports[`CheckboxGroup 组件 正确导出 CheckboxGroup 组件 defaultValue 参数 1`] = `
<div
  className="guree-radio-group"
>
  <span
    className="guree-checkbox guree-checkbox-checked"
    onClick={[Function]}
  >
    <span
      className="guree-checkbox-inner"
    >
      <svg
        className="guree-icon"
        style={
          Object {
            "fill": "#fff",
            "height": "10px",
            "width": "10px",
          }
        }
      >
        <use
          xlinkHref="#check"
        />
      </svg>
    </span>
    <span
      className="guree-checkbox-name"
    >
      1
    </span>
  </span>
  <span
    className="guree-checkbox"
    onClick={[Function]}
  >
    <span
      className="guree-checkbox-inner"
    />
    <span
      className="guree-checkbox-name"
    >
      2
    </span>
  </span>
  <span
    className="guree-checkbox"
    onClick={[Function]}
  >
    <span
      className="guree-checkbox-inner"
    />
    <span
      className="guree-checkbox-name"
    >
      3
    </span>
  </span>
</div>
`;

exports[`CheckboxGroup 组件 正确导出 CheckboxGroup 组件 value 参数 1`] = `
<div
  className="guree-radio-group"
>
  <span
    className="guree-checkbox guree-checkbox-checked"
    onClick={[Function]}
  >
    <span
      className="guree-checkbox-inner"
    >
      <svg
        className="guree-icon"
        style={
          Object {
            "fill": "#fff",
            "height": "10px",
            "width": "10px",
          }
        }
      >
        <use
          xlinkHref="#check"
        />
      </svg>
    </span>
    <span
      className="guree-checkbox-name"
    >
      1
    </span>
  </span>
  <span
    className="guree-checkbox"
    onClick={[Function]}
  >
    <span
      className="guree-checkbox-inner"
    />
    <span
      className="guree-checkbox-name"
    >
      2
    </span>
  </span>
  <span
    className="guree-checkbox"
    onClick={[Function]}
  >
    <span
      className="guree-checkbox-inner"
    />
    <span
      className="guree-checkbox-name"
    >
      3
    </span>
  </span>
</div>
`;
